OAK-10347 : Adding small util class for exporting a tree as flat file

stefan-egli opened the pull request.

stefan-egli committed:
OAK-10347 : Adding small util class for exporting a tree as flat file

stefan-egli commented:
@nfsantos, @thomasmueller , wdyt? can you please review?

stefan-egli committed:
unrelated testlet fix : allow slightly longer time for crash resolution

stefan-egli committed:
OAK-10347 : typo in log msg

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,109 @@
+        if (entryBatch.size() > 999) {
+            flush();
+        }
+        for (ChildNodeEntry e : ns.getChildNodeEntries()) {
```
If I understood correctly, this will be querying directly the local MongoDB instance. And this is doing a depth-first traversal of the tree rooted in NodeState given to the createFlatFileFor method. Is return of `getChildNodeEntries` deterministic? That is, can you be sure that the FFS file will always be the same for repositories with equivalent contents? Running a revision GC might change the order by which Mongo answers to some queries, even if the results are equivalent. If the order of results of `getChildNodeEntries` is not deterministic, then you could sort the child nodes before iterating over them.

stefan-egli replied:
Good point, I agree this should be addressed. Actually there's also the issue that `NodeStateEntryWriter` excludes the `:childOrder`.

I'm not sure what the best approach is yet. Might have to do some iteration over this code..

stefan-egli replied:
PS: also noticed that `nodeState.getProperties()` does not always return the same order - so the properties probably need to be sorted explicitly (in the `NodeStateEntryWriter`)

stefan-egli replied:
I have now added sorting of both children and properties (to the `SimpleFlatFileUtil` only, not by default to underlying classes). Also added an option to export the `:childOrder` property - this allows explicit veriication of this property (given the sorting does alphabetic sorting now anyway)

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,109 @@
+            // Here logic differs from NodeStateEntrySorter in sense that
+            // Holder line consist only of json and not 'path|json'
+            StateInBytesHolder h = new StateInBytesHolder(path, jsonText);
+            entryBatch.add(h);
```
Why not writing the entry directly to the file instead of having an intermediate step of storing it in the `entryBatch` array? The code is using a buffered writer, so I don't see what is the advantage of having a second layer of buffering with this entryBatch array.

stefan-egli replied:
+1, good point, I'll fix that.

nfsantos commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public static void createFlatFileFor(NodeState ns, File f) throws IOException {
+        final SimpleFlatFileUtil h = new SimpleFlatFileUtil(f);
+        log.info("createFlatFileFor : writing to {}", f.getCanonicalPath());
+        h.addEntryAndTraverseChildren(ns);
+        h.close();
```
nitpick: would be better to use try with resources to ensure the file is closed even when an error occurs.

stefan-egli replied:
+1, I'll fix that

nfsantos reviewed (approved):
LGTM, my only concern is whether the order of the entries in the FFS is deterministic.

stefan-egli committed:
OAK-10347 : incorporated review feedback

stefan-egli commented:
@nfsantos I tried to address your concerns (except the open ordering question)

stefan-egli committed:
unrelated testlet fix : allow even longer time for crash resolution

stefan-egli committed:
temporarily enabling console logger to analyse DocumentDiscoveryLiteServiceTest on sonarcloud

rishabhdaim commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,92 @@
+
+    public static void createFlatFileFor(NodeState ns, File f) throws IOException {
+        log.info("createFlatFileFor : writing to {}", f.getCanonicalPath());
+        try (FileWriter fw = new FileWriter(f);
+                BufferedWriter bw = new BufferedWriter(fw)) {
```
```suggestion
        try (BufferedWriter bw = Files.newBufferedReader(f.toPath())) {
```

rishabhdaim commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,92 @@
+    private final NodeStateEntryWriter entryWriter;
+    private long totalLines = 0;
+
+    private SimpleFlatFileUtil(BufferedWriter bw) throws IOException {
+        // blobStore is only used for deserialization - so pass null here:
+        entryWriter = new NodeStateEntryWriter(null);
+        this.bw = bw;
+    }
```
I would remove this constructor since it is a utility class and would pass the required parameters to methods directly.

stefan-egli replied:
I see. It's currently indeed a bit a mix between a utility class and an (writer) object - so actually the name might a bit off. But for now I don't see it doing other things than `createFlatFileFor`- so I tend to favour leaving it this way (but I agree, it's not the nicest class..)

stefan-egli committed:
temporarily ignoring successful tests to reduce log noise for debugging

stefan-egli committed:
and again, unrelated test debugging, more trace for ClusterNodeInfo

stefan-egli committed:
more temporary debug logging

stefan-egli committed:
unrelated : asserting clock is set correctly to default

stefan-egli committed:
unrelated buglet fix

stefan-egli committed:
unrelated : revert unrelated changes, except for actual test fix

stefan-egli committed:
Update oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java

Co-authored-by: Rishabh Kumar <rishabhdaim1991@gmail.com>

stefan-egli committed:
OAK-10347 : added properties sorting plus added some tests

stefan-egli committed:
OAK-10347 : added sorting to children by name

stefan-egli committed:
OAK-10347 : added support to export otherwise hidden :childOrder property in SimpleFlatFileUtil

stefan-egli committed:
OAK-10347 : handle one sonar smell, but ignore the other

stefan-egli commented:
@nfsantos , @thomasmueller , @rishabhdaim , I've addressed the review comments plus fixed an unrelated test case. I think the PR is now ready to merge. Can you please +1? Thx!

rishabhdaim commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/NodeStateEntryWriter.java:
```diff
@@ -77,21 +85,30 @@ public String toString(List<String> pathElements, String nodeStateAsJson) {
+
+    String asSortedJson(NodeState nodeState) {
+        return asJson(StreamSupport.stream(nodeState.getProperties().spliterator(), false)
+                .sorted((o1, o2) -> o1.getName().compareTo(o2.getName())));
```
```suggestion
                .sorted(Comparator.comparing(PropertyState::getName));
```

rishabhdaim commented on oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java:
```diff
@@ -0,0 +1,102 @@
+    private void addEntryAndTraverseChildren(NodeState ns) throws IOException {
+        addEntry(ns);
+        StreamSupport.stream(ns.getChildNodeEntries().spliterator(), false)
+                .sorted((o1, o2) -> o1.getName().compareTo(o2.getName())).forEach(e -> {
```
```suggestion
                .sorted(Comparator.comparing(ChildNodeEntry::getName)).forEach(e -> {
```

rishabhdaim committed:
Update oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/NodeStateEntryWriter.java

rishabhdaim committed:
Update oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/SimpleFlatFileUtil.java

rishabhdaim committed:
Update NodeStateEntryWriter.java

rishabhdaim committed:
Update SimpleFlatFileUtil.java

rishabhdaim commented:
@stefan-egli made minor changes directly to PR.

rishabhdaim committed:
fixed minor syntax issue

rishabhdaim committed:
Update SimpleFlatFileUtil.java

stefan-egli commented:
thx @rishabhdaim for those fixes!
